docs/rfc/rfc-001-storage-engine.rst

```diff
@@ -0,0 +1,179 @@
+the following high level goals: 
+
+- drop dependencies on storage engines that have a CGo dependency.
+
+- encapsulate data format and data storage from higher-level services
+  (e.g. reactors) within tendermint.
+
+- select a storage engine that does not incur any additional operational
+  complexity (e.g. database should be embedded.)
+
+- provide database semantics with sufficient ACID, snapshots, and
+  transactional support.
+
+Open Questions
+~~~~~~~~~~~~~~
+
+The following questions remain:
+
+- what kind of data-access concurrency does tendermint require?
+
+- would tendermint users SDK/etc. benefit from some shared database
+  infrastructure?
+  
+  - In earlier conversations it seemed as if the SDK has selected Badger and
+    RocksDB for their storage engines, and it might make sense to be able to
+    (optionally) pass a handle to a Badger instance between the libraries in
+    some cases.
+
+- what are typical data sizes, and what kinds of memory sizes can we expect
+  operators to be able to provide?
+
+- in addition to simple persistence, what kind of additional semantics would
+  tendermint like to enjoy (e.g. transactional semantics, unique constraints,
+  indexes, in-place-updates, etc.)?
+
+Decision Framework
+~~~~~~~~~~~~~~~~~~
+
+Given the constraint of removing the CGo dependency, the decision is between
+"badger" and "boltdb" (in the form of the etcd/CoreOS fork,) as low level. On
+top of this and somewhat orthogonally, we must also decide on the interface to
+the database and how the larger application will have to interact with the
+database layer. Users of the data layer shouldn't ever need to interact with
+raw byte slices from the database, and should mostly have the experience of
+interacting with Go-types.
+
+Badger is more consistently developed and has a broader featureset than
+Bolt. At the same time, Badger is likely more memory intensive and may have
+more overhead in terms of open file handles given it's model. At first glance,
+Badger is the obvious choice: it's actively developed and it has a lot of
+features that could be useful. Bolt is not without some benefits: it's stable
+and is maintained by the etcd folks, it's simpler model (single memory mapped
+file, etc,) may be easier to reason about.
+
+I propose that we consider the following specific questions about storage
+engines:
+
+- does Badger's evolving development, which may result in data file format
+  changes in the future, and could restrict our access to using the latest
+  version of the library between major upgrades, present a problem?
+
+- do we do we have goals/concerns about memory footprint that Badger may
+  prevent us from hitting, particularly as data sets grow over time?
+
+- what kind of additional tooling might we need/like to build (dump/restore,
+  etc.)?
+
+- do we want to run unit/integration tests against a data files on disk rather
+  than relying exclusively on the memory database?
```

Comment:
If there's a distinction in the characteristic of in-memory vs persistent databases and the integration tests are affected by such differences then I think that makes sense - but I think that's a very small subset of tests

Reply:
If we are really going to commit to a single implementation target, I think we shouldn't try to do anything fancy here: We should test the production code. One nice property of Badger is that it is almost entirely plain files on disk (it does create some advisory locks, but nothing fancy with device files or sockets as far as I know).

Reply:
I think our integration tests should certainly use the full on-disk storage engine. I would advocate that, where possible, we mock it out in the unit tests for the sake of simplicity and speed. For example, consensus tests that are not testing crash recovery do not require access to a real WAL and can and should use a mock (likely one that's implemented as a no-op).

Reply:
If we were talking about a heavy-weight database engine like Postgres, I'd agree. But the storage engines we've been considering are lightweight and fast: Fast enough that if we need to mock them out to make the tests go faster, we have probably already failed our performance goals in production.

So I don't think we can or should try to mock them in this case, though I do agree with the impulse more generally for "large" service dependencies.

Reply:
My concern here is mostly around the pain associated with adding and running more tests. If we're confident that the embedded DB can be fast enough that the slowdown of adding a new test is minimal and that we can easily write default setup/teardown code that makes adding DB access into a test very straightforward, then my concerns are largely unfounded.

Reply:
> My concern here is mostly around the pain associated with adding and running more tests. 

That's quite valid, but for these examples it wouldn't be any harder to add additional tests using the real production code than it would using a mock. 

I think it's best to avoid test-doubles unless using the production code is both (a) too complicated to do in a test environment (e.g., requires credentials, needs to spin up a separate service, etc.), and (b) not relevant enough to the code under test to be worth doing.

Here, I don't think we have either issue: Setting up a database is a couple lines of code, and the correct behaviour of the data store is relevant to anything tested against it. With the options we've been discussing, we could even just check in small databases literally as testdata, they're small and self-contained; creating new ones is just a few lines of code. 

Personally, I'd rather provide a support library (a la `httptest`) than have to deal with yet another mock that I can't be sure is actually implementing the correct semantics.

Reply:
Woah, this thread got out of control, and I think this was kind of a throw away point from my perspective. 

I mostly wanted to point out that except for the e2e tests, the database code is never run or used in our testing environment, and that's a flaw for a collection of reasons:
- because we should be able to run tests with more realistic situations when it make sense
- the current tests, because they're written against a hashmap, may miss some key behavior

This project will definitely entail building a new set of testing tools that will have to be fast/etc.